### PR TITLE
speedup for import astropy: "io.registry.get_formats"

### PR DESCRIPTION
**Pull request by MSeifert04**
Upon reviewing #5222 and #4598 I noticed that `io.registry.get_formats` is in the top50 of the cumulative time consumers when importing `astropy`.

Some measurements:

```
%prun from astropy.coordinates import SkyCoord
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
       68    0.021    0.000    1.213    0.018 registry.py:44(get_formats)

from astropy.io import registry
%load_ext line_profiler
%lprun -f registry.get_formats registry.get_formats()
Line #      Hits         Time  Per Hit   % Time  Line Contents
    83         1        47175  47175.0     74.0      format_table.sort(['Data class', 'Deprecated', 'Format'])

%timeit registry.get_formats()
100 loops, best of 3: 14.1 ms per loop
```
Compared with the same data with these changes
```
%prun from astropy.coordinates import SkyCoord
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
       68    0.019    0.000    0.350    0.005 registry.py:44(get_formats)

from astropy.io import registry
%load_ext line_profiler
%lprun -f registry.get_formats registry.get_formats()
Line #      Hits         Time  Per Hit   % Time  Line Contents
    80         1            5      5.0      0.0      data = None
    81         1            4      4.0      0.0      if rows:
    82         1          141    141.0      1.0          data = list(zip(*sorted(rows, key=itemgetter(0, 5, 1))))

%timeit registry.get_formats()
100 loops, best of 3: 4.07 ms per loop
```

This improves the import time of astropy by more than 10% on my computer. 

There is still room for lots of improvement, for example reducing the number of calls to this function, caching the result or improving the table-creation-time (which now takes 80% of the functions execution time). But this way it's a lot faster on my computer without having to do any fancy stuff. 

And I'm not sure if this actually works for everyone. Before I do more work here it would be good to know if this actually improves the import time for other people as well.

Btw: What makes the `Table.sort` so inefficient?

### CODE REVIEW

**Comment by MSeifert04** [comment]
The OSX Travis test will probably fail or take forever (see also https://www.traviscistatus.com/incidents/4mvp857qx8bw).

File: astropy/io/registry.py [hunk 1]
```diff
@@ -77,10 +77,11 @@ def get_formats(data_class=None):
         rows.append((format_class[1].__name__, format_class[0], has_read,
                      has_write, has_identify, deprecated))
 
-    data = list(zip(*rows)) if rows else None
+    data = None
+    if rows:
```
Comment (mhvk) [inline astropy/io/registry.py]: Maybe a silly stylistic comment, but I would do
```
if rows:
    data = ...
else:
    data = None
```
I think it shows the flow of the program more clearly.

Comment (MSeifert04) [inline astropy/io/registry.py]: I like silly stylistic comments. If someone else confirms that this is indeed a bottleneck when importing astropy I'll add an inline comment why this approach was chosen and use your suggestion.

Thank you!

**Comment by mhvk** [comment]
Excellent!  table sorting may be slower because the tables are collections of columns, not rows. So, it first constructs a sorting index, but then has to reshuffle every column individually using that index. Checking...
```
%prun -s cumulative from astropy.coordinates import SkyCoord

         1508566 function calls (1473098 primitive calls) in 1.639 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
       68    0.004    0.000    0.267    0.004 registry.py:43(get_formats)
       68    0.001    0.000    0.192    0.003 table.py:2224(sort)
       68    0.000    0.000    0.107    0.002 table.py:2183(argsort)
     408    0.001    0.000    0.028    0.000 {method 'take' of 'numpy.ndarray' objects}
     408    0.001    0.000    0.005    0.000 column.py:780(__setitem__)
```
So, not really true, the sorting itself is also long, in part because the to-be-sorted table is turned into a `recarray` first.

I'll at least partially blame our tendency to want to accept many things yet not duck-type:
```
   169185    0.034    0.000    0.060    0.000 {built-in method builtins.isinstance}
```
(really, nearly 170k calls just to import a module!?!?)

**Comment by mhvk** [comment]
So for me, it goes from  1.64 s in astropy 1.2.1 to 1.26 s in current master and  to 1.07 s. I think shaving off 15% of the import time is well worth this simple change!! (But am also happy to see that other parts have been sped up a bit...).

**Comment by mhvk** [comment]
I guess another question is why this formatted `Table` is created 68 times. I seems to be called from `_update__doc__`, i.e., it is just so that the list of available formats are available. Which of course should be available but given that the `__doc__` string will be accessed in only a small fraction of all cases where astropy is imported, it seems this would be a prime candidate for being a dynamically generated variable....  See #5264.

Obviously, if we had that, we wouldn't worry so much any more about the inefficiency here. What do you think? (Also cc @taldcroft, since git blames him for `_update__doc__`.)

**Comment by MSeifert04** [comment]
> it seems this would be a prime candidate for being a dynamically generated variable.... Obviously, if we had that, we wouldn't worry so much any more about the inefficiency here. What do you think?

If possible yes. It's not trivial because RTD relies on it and affiliated packages "might" as well. One would need an "it's finished now, let's create the table now" trigger and where would you place it? Or am I wrong?

**Comment by mhvk** [comment]
@MSeifert04 - I think it would just generate it whenever accessed, reflecting the state at that moment. But from #5264, it doesn't seem all that easy, so my sense would be to merge this as it does help immediately and is fine no matter what. @taldcroft - any objections?

**Comment by mhvk** [comment]
@MSeifert04 - I think this can just go in, but one more thing: could you add an entry in `CHANGES.rst`, under "other changes and additions" as a new `io.registry` entry? (I see that we put the `Angle` initialisation speed-up there too, and think this is a big enough improvement to make it worthwile.

You can add "[skip ci]" to the commit message so the whole travis set doesn't get run again.

**Comment by MSeifert04** [comment]
@mhvk I've included the Changelog entry.

**Comment by MSeifert04** [comment]
sorry, something went wrong when amending and it created 2 seperate commits. Should I squash them?

**Comment by mhvk** [comment]
@MSeifert04 - I find that with `github` one sometimes has to reload the page after rebase/amend to see the actual commits: at least, I see just the three relevant ones for this PR.  So, merging!

**Pull request merged**